firewall/inspection/portscan/detect.go

```diff
@@ -0,0 +1,357 @@
+	}
+}
+
+//updateScoreIgnoreBlockPrevOffender updates this 4 Values of the Struct
+//ipString needs to correspond to the key of the entry in the map ips
+//WARNING: This function maybe deletes the entry ipString from the Map ips. (look at the returncode)
+//return: still in map? (bool)
+func (d *ipData) updateScoreIgnoreBlockPrevOffender(ipString string) bool {
+	d.score -= intMin(int(time.Since(d.lastUpdated)/decreaseInterval), d.score)
+
+	if d.ignore {
+		if time.Since(d.lastSeen) > unignoreTime {
+			d.ignore = false
+		}
+	}
+
+	if d.previousOffender && time.Since(d.lastSeen) > undoSuspicionIdleTime {
+		d.previousOffender = false
+	}
+
+	if d.blocked && time.Since(d.lastSeen) > unblockIdleTime {
+		d.blocked = false
+		d.blockedPorts = []tcpUDPport{}
+
+		status.DeleteThreat(threadPrefix + ipString)
+	}
+
+	if !d.blocked && !d.ignore && !d.previousOffender && d.score == 0 {
+		delete(ips, ipString)
+		return false
+	}
+
+	d.lastUpdated = time.Now()
+	return true
+}
+
+// Destroy implements the destroy interface.
+func (d *Detector) Destroy() error {
+	return nil
+}
+
+// DetectorFactory is a primitive detection method that runs within the factory only.
+func DetectorFactory(conn *network.Connection, pkt packet.Packet) (network.Inspector, error) {
+	return &Detector{}, nil
+}
+
+// Register registers the encryption detection inspector with the inspection framework.
+func init() {
+	ips = make(map[string]*ipData)
+
+	//cleanup old Threads
+	threads, _ := status.GetThreats(threadPrefix)
+	for _, t := range threads {
+		status.DeleteThreat(t.ID)
+	}
+
+	module = modules.Register("portscan-detection", nil, start, nil, "base", "netenv")
+	module.Enable()
+}
+
+func updateWholeList() {
+	log.Debugf("Portscan detection: update list&cleanup")
+	for ip, entry := range ips {
+		//done inside the loop to give other goroutines time in between to access the list (and during that time block this task)
+		runOnlyOne.Lock()
+		defer runOnlyOne.Unlock()
+
+		if entry.updateScoreIgnoreBlockPrevOffender(ip) {
+			log.Debugf("%s: %v", ip, entry)
+		} else {
+			log.Debugf("Removed %s from the list", ip)
+		}
+	}
+	log.Debugf("Portscan detection: finished update list&cleanup")
+
+}
+
+func start() (err error) {
+	go delayedStart()
+
+	// Reload own IP List on Network change
+	err = module.RegisterEventHook(
+		"netenv",
+		"network changed",
+		"Reload List of own IPs on Network change for Portscan detection",
+		func(_ context.Context, _ interface{}) (err error) {
+			fillOwnIPs()
+
+			return
+		},
+	)
+
+	fillOwnIPs()
+
+	return
+}
+
+func delayedStart() {
+	time.Sleep(startAfter)
+
+	log.Debugf("starting Portscan detection")
+	err := inspection.RegisterInspector(&inspection.Registration{
+		Name:    "Portscan Detection",
+		Order:   0,
+		Factory: DetectorFactory,
+	})
+
+	if err != nil {
+		panic(err)
+	}
+
+	module.NewTask("portscan score update", func(ctx context.Context, task *modules.Task) (err error) {
+		updateWholeList()
+		return
```

Comment:
always (or at least prefer to) use values in returns: `return nil`. It's way easier to reason about them if you see the value and don't need to scroll up to the function definition (and scan the whole function body) to see that err is unassigned and that the default value is nil.